### ISSUE TITLE
ADODB connection to google cloud SQL from cloud run container?

### ISSUE THREAD

**Issue by joebordes** [ADOdb/ADOdb#641]
I can't seem to get the right combination of settings to connect to google cloud SQL from a run container. Has anyone accomplished that before and can share the magic encantation?

**Comment by mnewnham** [comment]
If you could provide us with a code snippet to explain what you are trying to achieve , and what connection errors you are receiving it would help a lot. There is no official support for this although its possible it could work

**Comment by loridacito** [comment]
Hello referring to what Joe is asking, we have a php application with a config file . Mysql configuration is as follows
$dbconfig['db_server'] = 'ip';
$dbconfig['db_port'] = ':3306';
$dbconfig['db_username'] = 'user';
$dbconfig['db_password'] = 'password';
$dbconfig['db_name'] = 'dbname';
$dbconfig['db_type'] = 'mysqli';

We are deploying this application to cloud run and we have a cloud SQL connection and database. We have added SQL cloud connection to our cloud run container from the interface.
We want to use the IP , user , password and database name of cloud SQL in our application as we always do in the VM instances, for example. By looking at the official documentation it seems that only PDO drive is recognized ([https://github.com/GoogleCloudPlatform/php-docs-samples/blob/master/cloud_sql/mysql/pdo/src/app.php]) and not ADOdb. Is there a way we can do it using ADOdb or is it hopeless?

**Comment by mnewnham** [comment]
1. Could you provide the actual connect string to the ADOdb connection please. In your above sample you provide a db_port option ":3306" which would not be necessary as this is the default mysqli option.
2. Could you run it with $db->debug enabled prior to issuing the connect() statement and post the results
3. Have you tried using the [ADOdb PDO Driver](http://adodb.org/dokuwiki/doku.php?id=v5:database:pdo) , if in fact the connection is limited to PDO

**Comment by loridacito** [comment]
> 1. Could you provide the actual connect string to the ADOdb connection please. In your above sample you provide a db_port option ":3306" which would not be necessary as this is the default mysqli option.
The connection string is the one we use everywhere else that includes the host (IP or domain), username , password and database name. You can find the port in the configuration file but it's there by default. Just in case.
> 2. Could you run it with $db->debug enabled prior to issuing the connect() statement and post the results
Warning: mysqli_query() expects parameter 1 to be mysqli, bool given in /var/www/html/include/adodb/drivers/adodb-mysqli.inc.php on line 788
Query: SET NAMES utf8 failed. Connection timed out
> 3. Have you tried using the [ADOdb PDO Driver](http://adodb.org/dokuwiki/doku.php?id=v5:database:pdo) , if in fact the connection is limited to PDO
We have an existing application and it would take too much effort to  change it using the adodb pdo drive

**Comment by mnewnham** [comment]
It looks like somewhere in your code you are trapping the connection failure and returning to the code, causing your **SET NAMES** statement to execute on an invalid database handle, but not actually displaying the database connection failure. It might help to test the connection using a standalone routine. For example here's a simple program to test the connection:

````php
error_reporting(E_ALL);
include '/dev/github/adodb/adodb.inc.php';

$db = ADOnewConnection('mysqli');
$db->debug = true;
$database = 'adodb';
$host     = '127.0.0.1';
$user     = 'root';
$password = 'banana';
$db->connect($host,$user,$password,$database);

print $db->isConnected();
````
If any of the connection parameters are invalid, then MySQL is going to return it's own error, e.g.
````
Could not connect : 
127.0.0.1: Access denied for user 'root'@'localhost' (using password: YES)
````

This is the error that I'm interested in

**Comment by loridacito** [comment]
The connection is not working because it seems Cloud Run doesn't recognize connections with Host (Public IP of Cloud SQL), user, password . It needs a DSN like this one mysql:dbname=dbtest;unix_socket=/cloudsql/testcloudsql:europe-west1:dbtest. But this kind of configuration can be used by pdo driver which is not possible in our case. We were wondering if there is a way to connect to Cloud SQL without using the pdo driver

**Comment by mnewnham** [comment]
I understand now. The  PHP mysqli driver itself does not accept DSN connection statements. Whilst the ADOdb driver does accept a DSN-like connection, it actually unpacks it in the background and feeds it to mysqli_real_connect in the correct non-DSN format.

You can define socket in the mysqli driver, so your connection would be: 

````php
$db = ADOnewConnection('mysqli');
db->socket = '/cloudsql/testcloudsql:europe-west1:dbtest';

$db->debug = true;
$database = 'dbtest';
$host     = 'ip';
$user     = 'someuser';
$password = 'somehost';
$db->connect($host,$user,$password,$database);
````




If you could use the PDO/mysqli driver , it is compatible with the mysqli driver in ADOdb, apart from a few data dictionary statements not supported by PDO, so unless you are using some very specific mysqli feature, you should be able to use that instead

**Comment by loridacito** [comment]
Thank you for your answer. I managed to add the socket as you said.  Unfortunately, the connection failed with a Connection Timeout because of the public IP of cloud SQL used as host. I tried to use localhost and it fails with a Connection Refused error. I have tried changing our driver to PDO but we receive errors in many queries. I believe it's due to the functions we use that are specific for mysqli.
